Sources/SwiftParser/Availability.swift

```diff
@@ -262,48 +262,96 @@ extension Parser {
+            break
+          }
+          let version = self.expectDecimalIntegerWithoutRecovery()
 
-        if i < maxComponentCount {
-          components.append(versionComponent)
-        } else {
-          trailingComponents.append(versionComponent)
-        }
-      }
+          let versionComponent = RawVersionComponentSyntax(period: period, number: version, arena: self.arena)
 
-      var unexpectedTrailingComponents: RawUnexpectedNodesSyntax?
+          if i < maxComponentCount {
+            components.append(versionComponent)
+          } else {
+            trailingComponents.append(versionComponent)
+          }
 
-      if !trailingComponents.isEmpty {
-        unexpectedTrailingComponents = RawUnexpectedNodesSyntax(elements: trailingComponents.compactMap { $0.as(RawSyntax.self) }, arena: self.arena)
-      }
+          if versionComponent.hasError {
+            let unexpectedComponents = components + trailingComponents
+            var unexpectedTokens = [RawSyntax(major)] + unexpectedComponents.map(RawSyntax.init)
+            if let (_, handle) = self.canRecoverTo(anyIn: VersionTupleSyntax.EndOfVersionTupleOptions.self),
+              handle.unexpectedTokens > 0
+            {
+              for _ in 0..<handle.unexpectedTokens {
+                unexpectedTokens.append(RawSyntax(self.consumeAnyToken()))
+              }
+            }
+            return RawVersionTupleSyntax(
+              RawUnexpectedNodesSyntax(unexpectedTokens, arena: self.arena),
+              major: self.missingToken(.integerLiteral, text: nil),
+              components: nil,
+              arena: self.arena
+            )
+          }
+        }
 
-      return RawVersionTupleSyntax(
-        major: major,
-        components: RawVersionComponentListSyntax(elements: components, arena: self.arena),
-        unexpectedTrailingComponents,
-        arena: self.arena
-      )
+        var unexpectedTrailingComponents: RawUnexpectedNodesSyntax?
+
+        if !trailingComponents.isEmpty {
+          unexpectedTrailingComponents = RawUnexpectedNodesSyntax(elements: trailingComponents.compactMap { $0.as(RawSyntax.self) }, arena: self.arena)
+        }
 
+        return RawVersionTupleSyntax(
+          major: major,
+          components: RawVersionComponentListSyntax(elements: components, arena: self.arena),
+          unexpectedTrailingComponents,
+          arena: self.arena
+        )
+      } else {
+        let unexpectedToken = self.eat(.floatingLiteral)
+        return RawVersionTupleSyntax(
+          RawUnexpectedNodesSyntax([unexpectedToken], arena: self.arena),
+          major: self.missingToken(.integerLiteral, text: nil),
+          components: nil,
+          arena: self.arena
+        )
+      }
     } else {
       let major = self.expectDecimalIntegerWithoutRecovery()
-      return RawVersionTupleSyntax(major: major, components: nil, arena: self.arena)
+      if major.isMissing {
```

Comment:
Similarly, in this context, if the integer literal major is missing, all tokens before the comma or right paren will be treated as unexpectedToken.
There are some duplicated codes. Would it be better to separate it into a function? 🤔